usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -5546,8 +6174,26 @@ CK_RV ep11tok_sign(STDLL_TokData_t * tokdata, SESSION * session,
     ep11_private_data_t *ep11_data = tokdata->private_data;
     CK_RV rc;
     SIGN_VERIFY_CONTEXT *ctx = &session->sign_ctx;
+    size_t keyblobsize = 0;
+    CK_BYTE *keyblob;
+    OBJECT *key_obj = NULL;
 
-    UNUSED(length_only);
+    rc = h_opaque_2_blob(tokdata, ctx->key, &keyblob, &keyblobsize, &key_obj,
+                          READ_LOCK);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s h_opaque_2_blob, rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+
+    rc = ep11tok_pkey_check(tokdata, session, key_obj, &ctx->mech);
+    switch (rc) {
+    case CKR_OK:
+        rc = sign_mgr_sign(tokdata, session, length_only, ctx, in_data,
+                           in_data_len, signature, sig_len);
+        goto done;
+    default:
+        goto done;
+    }
```

Comment:
There is something wrong here.
All switch cases end with got done. So it will never ever go to the EP11 path (below the switch).  But this can't be what you want. 

If in ep11tok_sign_init() you got CKR_FUNCTION_NOT_SUPPORTED and thus took the EP11 way, you also need to take the EP11 way here. 

This applies to all similar places, (verify, etc).

Please also run test without PKEY option enabled. I would suppose that those cases would currently not work anymore.

Reply:
Ah yes, we talked about the cases where an ep11 fallback is possible. Being in a series of sign/update or verify/update, or even just sign/verify, via pkey, we cannot fall back to ep11, because the init was done for pkey alone. 
ep11tok_sign_init() and ep11tok_verify_init() have a different switch construct, including a case for the ep11 fallback: 
```
    switch (rc) {
    case CKR_OK:
        rc = sign_mgr_init(tokdata, session, ctx, mech, recover_mode, key);
        free(ep11_sign_state);
        goto done;
    case CKR_FUNCTION_NOT_SUPPORTED:
        /* ep11 fallback */
        break;
    default:
        goto done;
    }

```
So yes, if an internal error occurs in sign, sign/update etc. we cannot recover. But I agree, as I just removed the middle case CKR_FUNCTION_NOT_SUPPORTED this looks a bit weird.  I can change this to like if (rc == CKR_OK) call sign/verify mgr else goto done.

Reply:
No you did not get my point. 
The point is that with your current code it NEVER ever goes to the EP11 path. 
It either goes to the PKEY path, or it goes to done (i.e. error). But you never go to the EP11 path, not even if PKEY is disabled, etc. A "if (rc == CKR_OK) call sign/verify mgr else goto done" would not change that.

I guess you simply need to add the CKR_FUNCTION_NOT_SUPPORTED case back in, doing a **break**, so that it falls through to the EP11 path, if PKEY is not supported/configured.

The only thing left is then if during init it gets CKR_OK, but in update it gets CKR_FUNCTION_NOT_SUPPORTED (or vice versa). Not sure if this can ever happen (in real life). This would lead to situations that probably cause other errors.....

Reply:
Yes, this was overoptimized. I added the CKR_FUNCTION_NOT_SUPPORTED case back in.
Done.